File: algopt/rebalancer/examples/website/cookbook/affinity_placement/weighted_affinity.py
#!/usr/bin/env python3
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

"""
Weighted Affinity Scores Variation

This variation demonstrates how to use different affinity strengths for
different relationships. Strong affinities are used for critical colocation
requirements, medium affinities for preferred placements, and weak affinities
for nice-to-have optimizations. This allows fine-grained control over placement
priorities.
"""

from algopt.rebalancer.interface.py_client.ProblemSolver import ProblemSolver
from rebalancer.interface.thrift.v2.ProblemSolver.thrift_types import (
    ConstraintSpecs,
    GoalSpecs,
    SolverSpecs,
)
from rebalancer.interface.thrift.v2.ProblemSpecs.thrift_types import (
    AssignmentAffinitiesSpec,
    AssignmentAffinity,
    CapacitySpec,
)
from rebalancer.interface.thrift.v2.SolverSpecs.thrift_types import (
    LocalSearchSolverSpec,
)


def place_with_weighted_affinities():
    """Place services with varying affinity strengths."""
    solver = ProblemSolver(
        service_name="microservice-placer", service_scope="production"
    )

    solver.setObjectName("service_instance")
    solver.setContainerName("host")

    # Setup hosts
    hosts = [f"host_{i}" for i in range(10)]
    assignment = {host: [] for host in hosts}

    # Create service instances
    services = {
        "api_gateway": 5,
        "cache": 10,
        "logger": 15,
    }

    all_instances = []
    for service_name, count in services.items():
        for i in range(count):
            instance_name = f"{service_name}_{i}"
            all_instances.append(instance_name)
            # Random initial assignment
            host_idx = len(all_instances) % len(hosts)
            assignment[hosts[host_idx]].append(instance_name)

    solver.setAssignment(assignment)

    # Add dimensions
    resource_usage = {inst: 4.0 for inst in all_instances}
    solver.addObjectDimension("cpu_usage", resource_usage)

    host_cpu_capacity = {host: 64.0 for host in hosts}
    solver.addContainerDimension("cpu_capacity", host_cpu_capacity)

    solver.addConstraint(
        ConstraintSpecs(
            capacitySpec=CapacitySpec(
                name="cpu-capacity",
                scope="host",
                dimension="cpu_usage",
            )
        )
    )

    # variation_start
    # Strong affinity (must colocate)
    critical_affinities = [
        AssignmentAffinity(
            object="api_gateway_0", scopeItemName="host_0", affinity=100.0
        )
    ]

    # Medium affinity (prefer colocate)
    moderate_affinities = [
        AssignmentAffinity(object="cache_0", scopeItemName="host_0", affinity=5.0),
        AssignmentAffinity(object="cache_1", scopeItemName="host_0", affinity=5.0),
    ]

    # Weak affinity (nice to have)
    weak_affinities = [
        AssignmentAffinity(object="logger_0", scopeItemName="host_0", affinity=1.0),
        AssignmentAffinity(object="logger_1", scopeItemName="host_0", affinity=1.0),
    ]
    # variation_end

    # Add affinity goals with different weights
    if critical_affinities:
        solver.addGoal(
            GoalSpecs(
                assignmentAffinitiesSpec=AssignmentAffinitiesSpec(
                    name="critical-affinity",
                    scope="host",
                    affinities=critical_affinities,
                )
            ),
            weight=10.0,  # Highest priority
        )

    if moderate_affinities:
        solver.addGoal(
            GoalSpecs(
                assignmentAffinitiesSpec=AssignmentAffinitiesSpec(
                    name="moderate-affinity",
                    scope="host",
                    affinities=moderate_affinities,
                )
            ),
            weight=3.0,  # Medium priority
        )

    if weak_affinities:
        solver.addGoal(
            GoalSpecs(
                assignmentAffinitiesSpec=AssignmentAffinitiesSpec(
                    name="weak-affinity",
                    scope="host",
                    affinities=weak_affinities,
                )
            ),
            weight=1.0,  # Low priority
        )

    # Use Local Search solver
    solver.addSolver(
        SolverSpecs(localSearchSolverSpec=LocalSearchSolverSpec(timeLimitMs=30000))
    )

    # Solve
    solution = solver.solve()

    print(f"Solution found in {solution.profile.solveTime}ms")
    print(f"Objective value: {solution.objectiveValue}")
    print(f"Instances moved: {solution.profile.moveCount}")

    # Analyze affinity satisfaction
    print("\nAffinity satisfaction:")
    print(
        f"  Critical: api_gateway_0 on host_0? {_check_affinity(solution.assignment, 'api_gateway_0', 'host_0')}"
    )
    print(
        f"  Moderate: cache_0 on host_0? {_check_affinity(solution.assignment, 'cache_0', 'host_0')}"
    )
    print(
        f"  Weak: logger_0 on host_0? {_check_affinity(solution.assignment, 'logger_0', 'host_0')}"
    )

    return solution


def _check_affinity(assignment, instance, host):
    """Check if instance is assigned to host."""
    return instance in assignment.get(host, [])


if __name__ == "__main__":
    place_with_weighted_affinities()

File: algopt/rebalancer/examples/website/cookbook/affinity_placement/weighted_affinity.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

/**
 * Weighted Affinity Scores Variation
 *
 * This variation demonstrates how to use different affinity strengths for
 * different relationships. Strong affinities are used for critical colocation
 * requirements, medium affinities for preferred placements, and weak affinities
 * for nice-to-have optimizations. This allows fine-grained control over
 * placement priorities.
 */

#include "algopt/rebalancer/interface/ProblemSolver.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/ProblemSpecs_types.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/SolverSpecs_types.h"

#include <fmt/format.h>
#include <folly/container/irange.h>
#include <folly/executors/CPUThreadPoolExecutor.h>
#include <folly/init/Init.h>

#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace facebook::rebalancer::interface;

int main(int argc, char* argv[]) {
  const folly::Init init(&argc, &argv);
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "microservice-placer", "production");

  solver.setObjectName("service_instance");
  solver.setContainerName("host");

  // Setup hosts
  std::vector<std::string> hosts;
  hosts.reserve(10);
  for (const auto i : folly::irange(10)) {
    hosts.push_back(fmt::format("host_{}", i));
  }

  std::map<std::string, std::vector<std::string>> assignment;
  for (const auto& host : hosts) {
    assignment[host] = {};
  }

  // Create service instances
  std::map<std::string, int> services = {
      {"api_gateway", 5},
      {"cache", 10},
      {"logger", 15},
  };

  std::vector<std::string> all_instances;
  for (const auto& [service_name, count] : services) {
    for (const auto i : folly::irange(count)) {
      std::string instance_name = fmt::format("{}_{}", service_name, i);
      all_instances.push_back(instance_name);
      // Random initial assignment
      int host_idx = all_instances.size() % hosts.size();
      assignment[hosts[host_idx]].push_back(instance_name);
    }
  }

  solver.setAssignment(assignment);

  // Add dimensions
  std::map<std::string, double> resource_usage;
  for (const auto& inst : all_instances) {
    resource_usage[inst] = 4.0;
  }
  solver.addObjectDimension("cpu_usage", resource_usage);

  std::map<std::string, double> host_cpu_capacity;
  for (const auto& host : hosts) {
    host_cpu_capacity[host] = 64.0;
  }
  solver.addContainerDimension("cpu_capacity", host_cpu_capacity);

  CapacitySpec cpuCapacity;
  cpuCapacity.name() = "cpu-capacity";
  cpuCapacity.scope() = "host";
  cpuCapacity.dimension() = "cpu_usage";
  solver.addConstraint(cpuCapacity);

  // variation_start
  // Strong affinity (must colocate)
  AssignmentAffinity critical;
  critical.objectName() = "api_gateway_0";
  critical.scopeItemName() = "host_0";
  critical.affinity() = 100.0;

  // Medium affinity (prefer colocate)
  AssignmentAffinity moderate1;
  moderate1.objectName() = "cache_0";
  moderate1.scopeItemName() = "host_0";
  moderate1.affinity() = 5.0;

  AssignmentAffinity moderate2;
  moderate2.objectName() = "cache_1";
  moderate2.scopeItemName() = "host_0";
  moderate2.affinity() = 5.0;

  // Weak affinity (nice to have)
  AssignmentAffinity weak1;
  weak1.objectName() = "logger_0";
  weak1.scopeItemName() = "host_0";
  weak1.affinity() = 1.0;

  AssignmentAffinity weak2;
  weak2.objectName() = "logger_1";
  weak2.scopeItemName() = "host_0";
  weak2.affinity() = 1.0;
  // variation_end

  // Add affinity goals with different weights
  AssignmentAffinitiesSpec criticalSpec;
  criticalSpec.name() = "critical-affinity";
  criticalSpec.scope() = "host";
  criticalSpec.affinities() = {critical};
  solver.addGoal(criticalSpec, 10.0); // Highest priority

  AssignmentAffinitiesSpec moderateSpec;
  moderateSpec.name() = "moderate-affinity";
  moderateSpec.scope() = "host";
  moderateSpec.affinities() = {moderate1, moderate2};
  solver.addGoal(moderateSpec, 3.0); // Medium priority

  AssignmentAffinitiesSpec weakSpec;
  weakSpec.name() = "weak-affinity";
  weakSpec.scope() = "host";
  weakSpec.affinities() = {weak1, weak2};
  solver.addGoal(weakSpec, 1.0); // Low priority

  // Use Local Search solver
  LocalSearchSolverSpec localSearch;
  localSearch.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  localSearch.solveTime() = 30000;
  solver.addSolver(localSearch);

  // Solve
  auto solution = solver.solve();

  std::cout << "AssignmentSolution found in " << "<solve time>" << "ms\n";
  std::cout << "Objective value: " << *solution.finalObjective()->value()
            << "\n";
  std::cout << "Instances moved: " << "<move count>" << "\n";

  return 0;
}
